protocol/x/affiliates/keeper/keeper.go

```diff
@@ -41,3 +49,154 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+) error {
+	affiliateReferredVolumePrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredVolumeKeyPrefix))
+	var referredVolume dtypes.SerializableInt
+	if !affiliateReferredVolumePrefixStore.Has([]byte(affiliateAddr)) {
+		referredVolume = dtypes.NewInt(0)
+	} else {
+		err := referredVolume.Unmarshal(affiliateReferredVolumePrefixStore.Get([]byte(affiliateAddr)))
+		if err != nil {
+			return err
+		}
+	}
+	referredVolumeBigInt := referredVolume.BigInt()
+	referredVolumeBigInt.Add(referredVolumeBigInt, referredVolumeFromBlock.BigInt())
+	updatedReferedVolume := dtypes.NewIntFromBigInt(referredVolumeBigInt)
+
+	updatedReferredVolumeBytes, err := updatedReferedVolume.Marshal()
+	if err != nil {
+		return errors.New("Error marshalling referred volume")
+	}
+	affiliateReferredVolumePrefixStore.Set([]byte(affiliateAddr), updatedReferredVolumeBytes)
+	return nil
+}
+
+func (k Keeper) GetReferredVolume(ctx sdk.Context, affiliateAddr string) (dtypes.SerializableInt, bool) {
+	affiliateReferredVolumePrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredVolumeKeyPrefix))
+	if !affiliateReferredVolumePrefixStore.Has([]byte(affiliateAddr)) {
+		return dtypes.NewInt(0), false
+	}
+	var referredVolume dtypes.SerializableInt
+	err := referredVolume.Unmarshal(affiliateReferredVolumePrefixStore.Get([]byte(affiliateAddr)))
+	if err != nil {
+		return dtypes.NewInt(0), false
+	}
+	return referredVolume, true
+}
+
+func (k Keeper) GetAllAffiliateTiers(ctx sdk.Context) (types.AffiliateTiers, error) {
+	store := ctx.KVStore(k.storeKey)
+	affiliateTiersBytes := store.Get([]byte(types.AffiliateTiersKey))
+
+	var affiliateTiers types.AffiliateTiers
+	if affiliateTiersBytes == nil {
+		return affiliateTiers, errors.New("Affiliate tiers not found")
+	}
+	err := k.cdc.Unmarshal(affiliateTiersBytes, &affiliateTiers)
+	if err != nil {
+		return affiliateTiers, err
+	}
+
+	return affiliateTiers, nil
+}
+
+func (k Keeper) GetTakerFeeShare(
+	ctx sdk.Context,
+	address string,
+) (
+	affiliateAddress string,
+	feeSharePpm uint32,
+	exists bool,
+	err error,
+) {
+	affiliateAddress, exists = k.GetReferredBy(ctx, address)
+	if !exists {
+		return "", 0, false, nil
+	}
+	_, feeSharePpm, err = k.GetTierForAffiliate(ctx, affiliateAddress)
+	if err != nil {
+		return "", 0, false, err
+	}
+	return affiliateAddress, feeSharePpm, true, nil
+}
+
+// Assumes that the affiliate tiers are sorted by level in ascending order.
+func (k Keeper) GetTierForAffiliate(
+	ctx sdk.Context,
+	affiliateAddr string,
+) (
+	tierLevel uint32,
+	feeSharePpm uint32,
+	err error) {
+	affiliateTiers, err := k.GetAllAffiliateTiers(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+	numTiers := uint32(len(affiliateTiers.GetTiers()))
+	maxTierLevel := numTiers - 1
+	currentTier := uint32(0)
+	referredVolume, exists := k.GetReferredVolume(ctx, affiliateAddr)
+
+	if !exists {
+		// If referred volume is not found, set it to 0.
+		referredVolume = dtypes.NewInt(0)
+	}
+
```

Comment:
Nit: assign `tiers := affiliateTiers.GetTiers()` to simplify below code

Reply:
Misunderstood this one, fixed!